### ISSUE TITLE
require Node >=14 for agoric-sdk

### ISSUE THREAD

**Issue by warner** [Agoric/agoric-sdk#1925]
Hm, it looks like the test failures are because Node 12.x does not expose `WeakRef` by default, but Node 14.x does. This might prompt us to increase our required Node.js version to 14, rather than instruct users to add whatever command-line option might enable weakrefs in 12 (if there even is such a thing.. I see `--harmony-weak-refs`, but it's labelled as "in progress").

_Originally posted by @warner in https://github.com/Agoric/agoric-sdk/pull/1924#issuecomment-716682755_

**Comment by warner** [comment]
@michaelfig noted the [Node.js Release Calendar](https://nodejs.org/en/about/releases/) which says Node 14 enters its LTS phase tomorrow (27-Oct-2020). He said v12 is not end-of-life until 30-Apr-2022, and it would be nice to not force people to upgrade to v14.

@erights says SES-shim would probably like to retain Node 12 compatibility. He suggested it would be easy to polyfill/stub `WeakRef` on platforms where it is not available (basically disabling GC), but was also disinclined to spend significant engineering time on support of an old release.

**Comment by warner** [comment]
Let's hold off on this for now, I'll look into a GC-disabling shim for Node v12.

**Comment by michaelfig** [comment]
At least Node v12.14.1 (which is our current minimum requirement) implements `--harmony-weak-refs` to enable WeakRef support.  That flag is compatible with v14.x.  So, I think it would be fairly straightforward to require that flag for all our top-level scripts and just use WeakRef unconditionally in SwingSet.

**Comment by warner** [comment]
In the meeting this morning, we decided to drop support for Node 12. @kriskowal floated the idea of dropping Node 14 as well (to make several `-r esm` / native-esm-support work better), but we agreed we can't do that until Node 16 reaches LTS.

**Comment by turadg** [comment]
> In the meeting this morning, we decided to drop support for Node 12. @kriskowal floated the idea of dropping Node 14 as well (to make several -r esm / native-esm-support work better), but we agreed we can't do that until Node 16 reaches LTS.

Now that [Node v16 is LTS](https://nodejs.org/ko/blog/release/v16.13.0/) shall we drop v14?

**Comment by michaelfig** [comment]
> Now that [Node v16 is LTS](https://nodejs.org/ko/blog/release/v16.13.0/) shall we drop v14?

I would prefer a "last 2 releases" policy.  Is node 18 a thing yet?

**Comment by turadg** [comment]
No Node 18 yet, though Node 17 is released so <=15 is older than "last 2 releases". https://nodejs.dev/download/ has a table of the releases and timing, showing that v16 is the "Active LTS" and v14 is what's called a "Maintenance LTS".

Why the preference for supporting older than the Active LTS?

**Comment by michaelfig** [comment]
> Why the preference for supporting older than the Active LTS?

@kriskowal has settled on supporting from v12 on the current endo, so I don't think his request implies we need to drop v14.  I only want to drop v14 support if it actually improves the SDK, and I'm not convinced of that yet.  Do you have reasons?

My hesitance is due to an aversion for breaking changes.

**Comment by turadg** [comment]
Makes sense. The reasons I thought it might be good to drop Node 14,
- ~ESM support mentioned above~ apparently irrelevant to this repo
- cut CI time (11 jobs listed below totalling 47min), though it wouldn't necessarily make the PR finish faster
- [features of v16](https://medium.com/the-node-js-collection/node-js-16-available-now-7f5099a97e70) such as,
  - Experimental support for Web Crypto API
  - Timers Promises API
  - npm 7
  - Stable `AbortController`
  - Stable Source Maps v3

<details>
<summary>PR checks that would be cut</summary>

```
Test all Packages / build (14.x) (pull_request) Successful in 4m
Test all Packages / test-quick (14.x) (pull_request) Successful in 5m
Test all Packages / test-quick2 (14.x) (pull_request) Successful in 4m
Test all Packages / test-solo (14.x) (pull_request) Successful in 4m
Test all Packages / test-cosmic-swingset (14.x) (pull_request) Successful in 6m
Test all Packages / test-swingset (14.x) (pull_request) Successful in 4m
Test all Packages / test-swingset2 (14.x) (pull_request) Successful in 3m
Test all Packages / test-swingset3 (14.x) (pull_request) Successful in 3m
Test all Packages / test-swingset4 (14.x) (pull_request) Successful in 4m
Test all Packages / test-zoe-unit (14.x) (pull_request) Successful in 5m
Test all Packages / test-zoe-swingset (14.x) (pull_request) Successful in 5m
```

</details>


I don't strongly advocate one way or the other. Mostly I'd like to see the decision rationale documented in the repo, which I'm happy to do.

**Comment by kriskowal** [comment]
I definitely intend to withdraw support for Node.js 12, though I’ve decoupled that from recent changes where I managed to keep support. Best reason to drop it is the experimental and different ESM implementation.
